File: telegram-formatter.js
/**
 * Identity Threat Alert Telegram Bot Formatter (Node.js)
 * Converts JSON payloads to SOC-friendly Telegram MarkdownV2 messages
 *
 * Usage:
 *   const { formatTelegramAlert } = require('./telegram-formatter');
 *   const message = formatTelegramAlert(payload);
 *
 * Or as CLI:
 *   node telegram-formatter.js payload.json
 */

/**
 * Escape Telegram MarkdownV2 special characters
 * @param {string|null} text - Text to escape
 * @returns {string} Escaped text safe for Telegram MarkdownV2
 */
function escapeTelegramMarkdown(text) {
  if (!text) return '';
  const specialChars = ['_', '*', '[', ']', '(', ')', '~', '`', '>', '#', '+', '-', '=', '|', '{', '}', '.', '!'];
  return String(text).split('').map(char => 
    specialChars.includes(char) ? `\\${char}` : char
  ).join('');
}

/**
 * Get severity emoji indicator
 * @param {string} severity - Severity level (low, medium, high, critical)
 * @returns {string} Emoji indicator
 */
function getSeverityEmoji(severity) {
  const emojiMap = {
    'low': '🟢',
    'medium': '🟡',
    'high': '🟠',
    'critical': '🔴'
  };
  return emojiMap[severity?.toLowerCase()] || '⚪';
}

/**
 * Format risk detections as bulleted list
 * @param {string[]} detections - Array of detection strings
 * @returns {string} Formatted bulleted list
 */
function formatRiskDetections(detections) {
  if (!detections || detections.length === 0) {
    return '• No specific indicators detected';
  }
  
  const formatted = detections.slice(0, 5).map(d => 
    `• ${escapeTelegramMarkdown(d)}`
  );
  
  if (detections.length > 5) {
    formatted.push(`• \\+${detections.length - 5} more indicators`);
  }
  
  return formatted.join('\n');
}

/**
 * Format investigation links as inline markdown
 * @param {object} links - Investigation links object
 * @returns {string} Formatted links
 */
function formatInvestigationLinks(links) {
  const linkLabels = {
    'signin_logs': 'Sign\\-in Logs',
    'oauth_apps': 'OAuth Apps',
    'inbox_rules': 'Inbox Rules',
    'sentinel_query': 'Sentinel Query',
    'user_profile': 'User Profile'
  };
  
  const formatted = [];
  for (const [key, label] of Object.entries(linkLabels)) {
    if (links[key]) {
      formatted.push(`[${label}](${links[key]})`);
    }
  }
  
  // Telegram MarkdownV2 requires '|' to be escaped as '\|'
  return formatted.join(' \\| ');
}

/**
 * Convert Identity Threat Alert JSON payload to Telegram MarkdownV2 message
 * @param {object} payload - Alert payload object
 * @returns {string} Formatted message string
 */
function formatTelegramAlert(payload) {
  const user = payload.user || {};
  const event = payload.event || {};
  const tech = payload.technical_indicators || {};
  const links = payload.investigation_links || {};
  const assigned = payload.assigned_to || {};
  const severity = payload.severity || 'medium';
  
  const severityEmoji = getSeverityEmoji(severity);
  
  // Build message parts
  const messageParts = [
    `${severityEmoji} *IDENTITY SECURITY ALERT*`,
    'Potential account compromise detected\\.',
    '',
    `👤 *User:* ${escapeTelegramMarkdown(user.email)}`,
    `🆔 *User ID:* ${escapeTelegramMarkdown(user.user_id)}`,
    `🏢 *Department:* ${escapeTelegramMarkdown(user.department)}`,
    `🔐 *Role:* ${escapeTelegramMarkdown(user.role)}`,
    '',
    '\\-\\-\\-',
    '',
    `⚠️ *Event Type:* ${escapeTelegramMarkdown(event.event_type)}`,
    escapeTelegramMarkdown(event.description),
    '',
    `🕒 *Timestamp:* ${escapeTelegramMarkdown(payload.timestamp_utc)}`,
    `🌐 *Source IP:* ${escapeTelegramMarkdown(event.source_ip)}`,
    `🏢 *ASN / Location:* ${escapeTelegramMarkdown(event.asn)} / ${escapeTelegramMarkdown(event.geo_location)}`,
    `📱 *Client App:* ${escapeTelegramMarkdown(event.client_app)}`,
    `🔐 *Auth Method:* ${escapeTelegramMarkdown(event.auth_method)}`,
    '',
    '\\-\\-\\-',
    '',
    '*🧩 Risk Indicators:*',
    formatRiskDetections(event.risk_detections),
    '',
    '\\-\\-\\-',
    '',
    '*🛑 Immediate Response Actions:*'
  ];
  
  // Add recommended actions
  const actions = payload.recommended_actions || [];
  for (let i = 0; i < Math.min(actions.length, 6); i++) {
    messageParts.push(`${i + 1}\\. ${escapeTelegramMarkdown(actions[i])}`);
  }
  
  if (actions.length > 6) {
    messageParts.push(`\\+ ${actions.length - 6} additional actions`);
  }
  
  messageParts.push(
    '',
    '\\-\\-\\-',
    '',
    '*🛠️ Investigation Links:*',
    formatInvestigationLinks(links),
    '',
    '\\-\\-\\-',
    '',
    `*📣 Severity:* ${severityEmoji} ${escapeTelegramMarkdown(severity.toUpperCase())}`,
    `*👮 Assigned To:* ${escapeTelegramMarkdown(assigned.analyst)}`
  );
  
  return messageParts.join('\n');
}

/**
 * CLI entry point
 */
if (require.main === module) {
  const fs = require('fs');
  const path = require('path');
  
  if (process.argv.length < 3) {
    console.error('Usage: node telegram-formatter.js <payload.json>');
    process.exit(1);
  }
  
  try {
    const filePath = process.argv[2];
    const payload = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    const message = formatTelegramAlert(payload);
    console.log(message);
  } catch (err) {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  }
}

module.exports = {
  formatTelegramAlert,
  escapeTelegramMarkdown,
  getSeverityEmoji
};

File: telegram_formatter.py
#!/usr/bin/env python3
"""
Identity Threat Alert Telegram Bot Formatter
Converts JSON payloads to SOC-friendly Telegram messages

Usage:
  from telegram_formatter import format_telegram_alert
  message = format_telegram_alert(payload_dict)
  
Or as CLI:
  python telegram_formatter.py payload.json
"""

import json
import sys
from typing import Dict, Any, Optional


def escape_telegram_markdown(text: Optional[str]) -> str:
    """
    Escape Telegram MarkdownV2 special characters.
    MarkdownV2 requires escaping: _ * [ ] ( ) ~ ` > # + - = | { } . !
    """
    if text is None:
        return ""
    
    text = str(text)
    special_chars = ['_', '*', '[', ']', '(', ')', '~', '`', '>', '#', '+', '-', '=', '|', '{', '}', '.', '!']
    for char in special_chars:
        text = text.replace(char, f'\\{char}')
    return text


def get_severity_emoji(severity: str) -> str:
    """Map severity level to emoji indicator"""
    emoji_map = {
        'low': '🟢',
        'medium': '🟡',
        'high': '🟠',
        'critical': '🔴'
    }
    return emoji_map.get(severity.lower(), '⚪')


def format_risk_detections(detections: list) -> str:
    """Format risk detections as a bulleted list"""
    if not detections:
        return "• No specific indicators detected"
    
    formatted = []
    for detection in detections[:5]:  # Limit to 5 for Telegram readability
        formatted.append(f"• {escape_telegram_markdown(detection)}")
    
    if len(detections) > 5:
        formatted.append(f"• \\+{len(detections) - 5} more indicators")
    
    return '\n'.join(formatted)


def format_investigation_links(links: Dict[str, str]) -> str:
    """Format investigation links as inline buttons"""
    formatted = []
    
    link_labels = {
        'signin_logs': 'Sign\\-in Logs',
        'oauth_apps': 'OAuth Apps',
        'inbox_rules': 'Inbox Rules',
        'sentinel_query': 'Sentinel Query',
        'user_profile': 'User Profile'
    }
    
    for key, label in link_labels.items():
        if key in links and links[key]:
            # Format as markdown link: [text](url)
            formatted.append(f"[{label}]({links[key]})")
    
    return ' | '.join(formatted)


def format_telegram_alert(payload: Dict[str, Any]) -> str:
    """
    Convert Identity Threat Alert JSON payload to Telegram MarkdownV2 message.
    
    Args:
        payload: Dictionary containing the alert payload
        
    Returns:
        Formatted message string safe for Telegram MarkdownV2
    """
    user = payload.get('user', {})
    event = payload.get('event', {})
    tech = payload.get('technical_indicators', {})
    links = payload.get('investigation_links', {})
    assigned = payload.get('assigned_to', {})
    severity = payload.get('severity', 'medium')
    
    severity_emoji = get_severity_emoji(severity)
    
    # Build message parts
    message_parts = [
        f"{severity_emoji} *IDENTITY SECURITY ALERT*",
        "Potential account compromise detected\\.",
        "",
        f"👤 *User:* {escape_telegram_markdown(user.get('email', 'N/A'))}",
        f"🆔 *User ID:* {escape_telegram_markdown(user.get('user_id', 'N/A'))}",
        f"🏢 *Department:* {escape_telegram_markdown(user.get('department', 'N/A'))}",
        f"🔐 *Role:* {escape_telegram_markdown(user.get('role', 'N/A'))}",
        "",
        "\\-\\-\\-",
        "",
        f"⚠️ *Event Type:* {escape_telegram_markdown(event.get('event_type', 'N/A'))}",
        f"{escape_telegram_markdown(event.get('description', 'N/A'))}",
        "",
        f"🕒 *Timestamp:* {escape_telegram_markdown(payload.get('timestamp_utc', 'N/A'))}",
        f"🌐 *Source IP:* {escape_telegram_markdown(event.get('source_ip', 'N/A'))}",
        f"🏢 *ASN / Location:* {escape_telegram_markdown(event.get('asn', 'N/A'))} / {escape_telegram_markdown(event.get('geo_location', 'N/A'))}",
        f"📱 *Client App:* {escape_telegram_markdown(event.get('client_app', 'N/A'))}",
        f"🔐 *Auth Method:* {escape_telegram_markdown(event.get('auth_method', 'N/A'))}",
        "",
        "\\-\\-\\-",
        "",
        "*🧩 Risk Indicators:*",
        format_risk_detections(event.get('risk_detections', [])),
        "",
        "\\-\\-\\-",
        "",
        "*🛑 Immediate Response Actions:*",
    ]
    
    # Add recommended actions with numbering
    actions = payload.get('recommended_actions', [])
    for i, action in enumerate(actions[:6], 1):  # Limit to 6 actions
        message_parts.append(f"{i}\\. {escape_telegram_markdown(action)}")
    
    if len(actions) > 6:
        message_parts.append(f"\\+ {len(actions) - 6} additional actions")
    
    message_parts.extend([
        "",
        "\\-\\-\\-",
        "",
        "*🛠️ Investigation Links:*",
        format_investigation_links(links),
        "",
        "\\-\\-\\-",
        "",
        f"*📣 Severity:* {severity_emoji} {escape_telegram_markdown(severity.upper())}",
        f"*👮 Assigned To:* {escape_telegram_markdown(assigned.get('analyst', 'N/A'))}"
    ])
    
    return '\n'.join(message_parts)


def main():
    """CLI entry point"""
    if len(sys.argv) < 2:
        print("Usage: python telegram_formatter.py <payload.json>")
        sys.exit(1)
    
    try:
        with open(sys.argv[1], 'r') as f:
            payload = json.load(f)
        
        message = format_telegram_alert(payload)
        print(message)
        
    except FileNotFoundError:
        print(f"Error: File '{sys.argv[1]}' not found")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON - {e}")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == '__main__':
    main()
